### PR TITLE
feat(governance): governance contract

### LINKED ISSUE #50: Make a governance contract
The governance contract should allow token holders (of a MiniMe token) to vote on proposals. A proposal is a call to an other smart contract with the amount of weis to transfer and the input data (which can be used to call a function with some parameters).
When a proposal is submitted, a fork of the MiniMe token is done and the forked token used as transferable vote tokens (to allow liquid voting).
For now we assume that everyone can submit a proposal but in the future proposal submission will be limited.

### PR DESCRIPTION
**Pull request by ferittuncer**
Fixes #50

### CODE REVIEW

**Comment by ferittuncer** [comment]
Tests and comments to do.

**Comment by ferittuncer** [comment]
Calling functions with struct parameters (https://github.com/ethereum/web3.js/issues/1241) got support in the latest web3 release `1.0.0-beta.36`. However `ganache-core` still depends on `1.0.0-beta.35`. 

So, for now, I will change the contract a bit to be able to call and test it from Javascript.

**Comment by ferittuncer** [comment]
KlerosLiquid knows it's governor(Governance contract) and Governance knows it's token controller(KlerosLiquid contract).

File: contracts/Governance.sol [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+        quorumDivideTime = _quorumDivideTime;
+        votingTime = _votingTime;
+    }
+
+
+    // ****************************** //
+    // *          Modifiers         * //
+    // ****************************** //
+
+    modifier onlyItself() {
```
Comment (satello) [inline contracts/Governance.sol]: Why not just use `internal` or `private`?

Comment (ferittuncer) [reply contracts/Governance.sol]: Correct me if I'm wrong:

When internal, a contract which inherits Governance (which is unauthorized) contract can call it.
When private, a contract which inherits Governance contract won't be able to reach that function, has to redefine.

Comment (satello) [reply contracts/Governance.sol]: So what is wrong with private? If the inheriting contract is unauthorized to call it, why does it need to be able to reach it?

Comment (ferittuncer) [reply contracts/Governance.sol]: As far as I know subclasses do not inherit private members, so a subclass will have to define those functions again. Shortly, extra work for users of this class.

Comment (ferittuncer) [reply contracts/Governance.sol]: No, you are right, `internal` does the job, `onlyItself` is unnecessary. Gonna fix this tomorrow. Thanks.

File: contracts/Governance.sol [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    modifier onlyWhenProposalIsDecided(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.Decided, "Only when proposal in Decided state.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriod(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp <= proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period.");
```
Comment (epiqueras) [inline contracts/Governance.sol]: Time periods are usually `[start, end)`.
So `block.timestamp < proposals[_id].whenPutToVote + currentVotingTime`.
But `now - proposals[_id].whenPutToVote < currentVotingTime` reads nicer.

File: contracts/Governance.sol [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+    modifier onlyWhenQuorumReached(bytes32 _id) {
+        require(proposals[_id].quorumToken.balanceOf(supportDeposit) >= getRequiredQuorum(_id), "Only when propsal has quorum.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriodPassed(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp > proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period passed.");
```
Comment (epiqueras) [inline contracts/Governance.sol]: now - proposals[_id].whenPutToVote >= currentVotingTime

File: contracts/Governance.sol [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+    event ProposalPutToSupport(bytes32 indexed _id);
+
+    event ProposalPutToVote(bytes32 indexed _id);
+
+    event ProposalDecided(bytes32 indexed _id, bool _approved);
+
+    event ProposalExecuted(bytes32 indexed _id);
+
+    event ProposalCreated(bytes32 indexed _id, address _destination, uint _amount, bytes _data, string _uriDescription, bytes32 _descriptionHash, string _uriArguments, bytes32 _argumentsHash);
```
Comment (epiqueras) [inline contracts/Governance.sol]: It's better to define events in order, so this one should go first.

Most of the time, you won't need all those params. You'll just fetch the events and then request details for the ones you care about. So I think `indexed _id` and `indexed _destination` is sufficient.
They cover the two use cases:
- Find out when a proposal was created from `_id`.
- Find all proposals for a `_destination`.

File: contracts/Governance.sol [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+    function createProposal(bytes32 _id, address _destination, uint _amount, bytes _data, string _uriDescription, string _uriArguments) public {
+        require(proposals[_id].destination == address(0), "There is already a proposal with this id");
+
+        proposals[_id].destination = _destination;
+        proposals[_id].amount = _amount;
+        proposals[_id].data = _data;
+        proposals[_id].uriDescription = _uriDescription;
+        proposals[_id].uriArguments = _uriArguments;
+
+        proposals[_id].descriptionHash = keccak256(proposals[_id].uriDescription);
```
Comment (epiqueras) [inline contracts/Governance.sol]: These are supposed to be the hashes of the file contents, not the URIs, so they should be params to the function.

File: contracts/Governance.sol [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+
+        emit ProposalCreated(_id, _destination, _amount, _data, _uriDescription, proposals[_id].descriptionHash, _uriArguments, proposals[_id].argumentsHash);
+    }
+
+
+    /** @dev Request registering a proposal to the proposal list
+     *  @param _id ID of a proposal
+     */
+    function requestRegisteringProposal(bytes32 _id) public payable onlyWhenProposalIsNew(_id) {
+        proposalList.requestRegistration.value(msg.value)(_id);
```
Comment (epiqueras) [inline contracts/Governance.sol]: Why not do this on proposal creation?

File: contracts/Governance.sol [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+        address cloneToken = pinakion.createCloneToken({_cloneTokenName: voteTokenName, _cloneDecimalUnits: DECIMALS, _cloneTokenSymbol: voteTokenSymbol, _snapshotBlock: block.number, _transfersEnabled: true});
+        proposal.voteToken = MiniMeToken(cloneToken);
+        proposal.voteToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToVote;
+
+        emit ProposalPutToVote(_id);
+
+        resetSettings(); // Reset global variables as quorum has reached
```
Comment (epiqueras) [inline contracts/Governance.sol]: No need to make a function if only called once.

File: contracts/Governance.sol [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+
+    /** @dev Ends a voting, moves proposal to decided state, sets the decision
+     *  @param _id ID of a proposal
+     */
+    function finalizeVoting(bytes32 _id) public onlyWhenProposalPutToVote(_id) onlyWhenVotingPeriodPassed(_id) {
```
Comment (epiqueras) [inline contracts/Governance.sol]: The two modifiers are redundant, you might have other cases like these in other functions too.

File: contracts/Governance.sol [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    }
+
+
+    /** @dev General purpose call function for executing a proposal UNTRUSTED
+     *  @param _id ID of a proposal
+     */
+    function executeProposal(bytes32 _id) public onlyWhenApproved(_id) {
+        Proposal storage proposal = proposals[_id];
+
+        proposal.destination.call.value(proposal.amount)(proposal.data); // solium-disable-line security/no-call-value
```
Comment (epiqueras) [inline contracts/Governance.sol]: Where was the value for `proposal.amount` provided? We need to require it somewhere along the flow.

File: test/governance.js [hunk 10]
```diff
@@ -0,0 +1,691 @@
+/* eslint-disable no-undef */ // Avoid the linter considering truffle elements as undef.
+
+const BigNumber = web3.BigNumber
+const {
+  expectThrow
+} = require('openzeppelin-solidity/test/helpers/expectThrow')
```
Comment (epiqueras) [inline test/governance.js]: We have these in `kleros-interaction`.

File: contracts/Governance.sol (hunk 9 above)
Comment (epiqueras) [reply contracts/Governance.sol]: Or will we top off the contract's balance every so often?

**Comment by ferittuncer** [comment]
### Things I found while making changes for Enriques review: 

Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:


1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.

2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.

File: test/governance.js (hunk 10 above)
Comment (ferittuncer) [reply test/governance.js]: Hmm, why don't we just let OpenZeppelin maintain them for us?

File: contracts/Governance.sol (hunk 9 above)
Comment (ferittuncer) [reply contracts/Governance.sol]: Three options come to my mind:

1. Make proposer pay when creating a proposal.
2. Let the contract pay as you said.
3. Collect the required amount from voters who approved. This is complicated much more complicated than the first and the second option.

What do you think? @satello @n1c01a5 @clesaege

File: contracts/Governance.sol (hunk 5 above)
Comment (ferittuncer) [reply contracts/Governance.sol]: I couldn't understand this, how to hash the content as we don't keep it on contract?

File: contracts/Governance.sol [hunk 11]
```diff
@@ -0,0 +1,256 @@
...
+    // *    Governance Mechanism    * //
+    // ****************************** //
+
+    /** @dev Creates a proposal, adds to proposals and requests registering to proposalList
+     *  @param _id ID of the proposalList
+     *  @param _destination Destination contract of the execution
+     *  @param _amount Value of the execution
+     *  @param _data Data of the execution
+     *  @param _uriDescription URI of the description of the proposal
+     *  @param _uriArguments URI of the arguments of the proposal
```
Comment (satello) [inline contracts/Governance.sol]: I would swap the wording in these two names (_uriArguments -> _argumentsURI and _uriDescription -> _descriptionURI)

File: contracts/Governance.sol (hunk 5 above)
Comment (satello) [reply contracts/Governance.sol]: The submitter hashes it off chain and submits it as a param to the contract. The contract can't verify but whoever looks up the proposal can

File: contracts/Governance.sol [hunk 12]
```diff
@@ -0,0 +1,256 @@
...
+
+        address cloneToken = pinakion.createCloneToken({_cloneTokenName: voteTokenName, _cloneDecimalUnits: DECIMALS, _cloneTokenSymbol: voteTokenSymbol, _snapshotBlock: block.number, _transfersEnabled: true});
+        proposal.voteToken = MiniMeToken(cloneToken);
+        proposal.voteToken.changeController(tokenController);
+
+        proposal.state = ProposalState.PutToVote;
+
+        emit ProposalPutToVote(_id);
+
+        lastTimeQuorumReached = block.timestamp; // Necessary when calculating required quorum as it is halved periodically.
```
Comment (satello) [inline contracts/Governance.sol]: What happens when more than one proposal is being voted on at the same time? Does resetting these global variables effect the first vote?

File: contracts/Governance.sol [hunk 13]
```diff
@@ -0,0 +1,256 @@
...
+        emit ProposalDecided(_id, proposals[_id].approved);
+    }
+
+
+    /** @dev General purpose call function for executing a proposal UNTRUSTED
+     *  @param _id ID of a proposal
+     */
+    function executeProposal(bytes32 _id) public {
+        require(proposals[_id].state == ProposalState.Decided, "Only when proposal in Decided state.");
+        require(proposals[_id].approved, "Only when proposal is approved.");
```
Comment (satello) [inline contracts/Governance.sol]: I would change the wording on all of these exception messages from "Only when..." to "... is not ..." or "... must be ....". Like `Proposal is not approved` and `Proposal must be in Decided state`

File: contracts/Governance.sol (hunk 12 above)
Comment (ferittuncer) [reply contracts/Governance.sol]: This is about quorum (support) voting. When a proposal reaches the quorum, the timer resets.

Quorum requirement has a half-life, if in x seconds no proposal reaches quorum, requirement gets halved.

- Let's say there are two proposals who have %35 support at the moment. 
- Base quorum requirement is %60 but it got halved one time, so it's %30 now. 

In this case there is a race, first call to `putProposalToVote()` wins and resets quorum requirement to %60.

Comment (ferittuncer) [reply contracts/Governance.sol]: Hmm, I understand what you mean, yes the first one will change the global variables and affect the second one.

Should we keep it constant after proposal creation? If so I can keep those variables in the struct and initialize them when created.

File: contracts/Governance.sol (hunk 9 above)
Comment (epiqueras) [reply contracts/Governance.sol]: I think 2 makes more sense. One party should not have to fund the organization's decisions.

File: contracts/Governance.sol (hunk 12 above)
Comment (epiqueras) [reply contracts/Governance.sol]: I think it's OK that it affects the second one. It should reflect the latest activity levels of members.

File: test/governance.js (hunk 10 above)
Comment (epiqueras) [reply test/governance.js]: Then we should change it everywhere else as well. I'll make an issue.

File: package.json [hunk 14]
```diff
@@ -43,6 +43,7 @@
   },
   "dependencies": {
     "kleros-interaction": "^0.0.25",
-    "minimetoken": "^0.2.0"
+    "minimetoken": "^0.2.0",
+    "openzeppelin-solidity": "^1.12.0"
```
Comment (epiqueras) [inline package.json]: It's a dev dependency.

**Comment by epiqueras** [comment]
> ### Things I found while making changes for Enriques review:
> Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> 
> 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.

https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361

**Comment by ferittuncer** [comment]
> > ### Things I found while making changes for Enriques review:
> > Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> > 
> > 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> > 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.
> 
> https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361

Yeah, I know that we can change the controller and we change it too. Token controller will be KlerosLiquid contract. That's the problem, only KlerosLiquid can disable transfers.

**Comment by ferittuncer** [comment]
> > > ### Things I found while making changes for Enriques review:
> > > Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> > > 
> > > 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> > > 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.
> > 
> > 
> > https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361
> 
> Yeah, I know that we can change the controller and we change it too. Token controller will be KlerosLiquid contract. That's the problem, only KlerosLiquid can disable transfers.

No, you mean Governance is the controller, now I get it. But as I said, we change the controller when we clone. This is what Athena specification wants.

**Comment by epiqueras** [comment]
> > > > ### Things I found while making changes for Enriques review:
> > > > Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> > > > 
> > > > 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> > > > 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.
> > > 
> > > 
> > > https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361
> > 
> > 
> > Yeah, I know that we can change the controller and we change it too. Token controller will be KlerosLiquid contract. That's the problem, only KlerosLiquid can disable transfers.
> 
> No, you mean Governance is the controller, now I get it. But as I said, we change the controller when we clone. This is what Athena specification wants. @epiqueras

From the spec:
> For now, this token controller will just be 0x0. Once the liquid voting contract is developed, it will be the liquid voting contract.

File: contracts/Governance.sol (hunk 12 above)
Comment (epiqueras) [reply contracts/Governance.sol]: Actually, from the spec:
> The quorum for this proposal is saved such that quorum modifications do not affect it.

**Comment by ferittuncer** [comment]
> > > > > ### Things I found while making changes for Enriques review:
> > > > > Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> > > > > 
> > > > > 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> > > > > 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.
> > > > 
> > > > 
> > > > https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361
> > > 
> > > 
> > > Yeah, I know that we can change the controller and we change it too. Token controller will be KlerosLiquid contract. That's the problem, only KlerosLiquid can disable transfers.
> > 
> > 
> > No, you mean Governance is the controller, now I get it. But as I said, we change the controller when we clone. This is what Athena specification wants. @epiqueras
> 
> From the spec:
> 
> > For now, this token controller will just be 0x0. Once the liquid voting contract is developed, it will be the liquid voting contract.

I guess liquid voting contract is not KlerosLiquid, right?

**Comment by epiqueras** [comment]
> > > > > > ### Things I found while making changes for Enriques review:
> > > > > > Voting is done by sending tokens to a predefined address. So the contract can start the voting process by cloning tokens but can't end it, because Governance is not the token controller. When we call `finalizeVoting()` Governance calculates the outcome and writes it to storage. Two minor issues here:
> > > > > > 
> > > > > > 1. When `finalizeVoting()` gets called, the contract sets the proposal as `Decided` and sets `approved` variable. But voting is still open. So if someone after this period keeps voting, it will corrupt final vote counts, but not change the outcome. We can avoid this by keeping another record when `finalizeVoting()` gets called.
> > > > > > 2. Voting continues unless someone calls `finalizeVoting()`. So in effect `votingTime` is the minimum and infinity is the maximum. To end it transfers can be disabled by the token controller. We can avoid this by making the voting process like this: Voter sends tokens to Governance contract address, Governance reflects transferred tokens as votes to the counter. And this function gets disabled after voting periods ends. This also fixes issue 1.
> > > > > 
> > > > > 
> > > > > https://github.com/Giveth/minime/blob/ea04d950eea153a04c51fa510b068b9dded390cb/contracts/MiniMeToken.sol#L361
> > > > 
> > > > 
> > > > Yeah, I know that we can change the controller and we change it too. Token controller will be KlerosLiquid contract. That's the problem, only KlerosLiquid can disable transfers.
> > > 
> > > 
> > > No, you mean Governance is the controller, now I get it. But as I said, we change the controller when we clone. This is what Athena specification wants. @epiqueras
> > 
> > 
> > From the spec:
> > > For now, this token controller will just be 0x0. Once the liquid voting contract is developed, it will be the liquid voting contract.
> 
> I guess liquid voting contract is not KlerosLiquid, right?

LiquidVoting is the contract that implements delegation logic.
Which can have a proxy for the governance contract to define action callbacks and/or block transfers.
So it's OK if you leave Governance as the token controller for now.

File: contracts/Governance.sol (hunk 12 above)
Comment (ferittuncer) [reply contracts/Governance.sol]: > Actually, from the spec:
> 
> > The quorum for this proposal is saved such that quorum modifications do not affect it.

So we should keep it constant after proposal creation, OK.

**Comment by ferittuncer** [comment]
Summary on what's left unfixed:

1. The quorum for this proposal is saved such that quorum modifications do not affect it.
2. Governance is the token controller for now. KlerosLiquid is not a liquid voting contract.

**Comment by ferittuncer** [comment]
The last issue: https://github.com/kleros/kleros/pull/87#discussion_r219333467

**Pull request merged**